clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -7099,6 +7099,17 @@ def warn_format_nonliteral : Warning<
   "format string is not a string literal">,
   InGroup<FormatNonLiteral>, DefaultIgnore;
 
+def err_non_static_member_use_not_allowed : Error<
+  "use of non-static member variable %0 is not allowed">;
+def err_not_publicly_accessible: Error<
+  "member variable %0 not publicly accessible">;
+def err_array_of_device_global_not_allowed : Error<
+  "array of device_global %0 is not allowed">;
+def err_shadow_variable_within_same_namespace: Error<
+  "shadow variable %0 not allowed withing the same enclosing namespace scope">;
+def err_namespace_name_shadows_namespace_containing_device_global : Error<
+  "not allowed: namespace name shadows %0 namespace which contains device_global">;
+
```

Comment:
Yeah, so as I suggested, let's focus on the functional part described by the design doc first. I'll take a closer look tomorrow on how we can implement diagnosing. But now, I'm not even sure how to do that, since the attribute applies to the class, but the restrictions are applied to a concrete object of that class.

Reply:
Sema::ActOnVariableDeclarator in SemaDecl.cpp looks promising